File: testing/iOS/StyleConverterTest/StyleEngine/typography/font/FontFamilyExtractor.swift
//
//  FontFamilyExtractor.swift
//  StyleEngine/typography/font — Phase 6.
//
//  The IR blob for `font-family` is either an array of
//  `{ "name": "..." }` / `{ "keyword": "serif" }` objects or a single
//  string (shortcut when there was one unquoted family name).
//

import Foundation

enum FontFamilyProperty { static let name = "FontFamily" }

enum FontFamilyExtractor {
    static func extract(from properties: [IRProperty]) -> FontFamilyConfig? {
        // Capture the last FontFamily wins (CSS cascade).
        var cfg = FontFamilyConfig()
        var touched = false
        for prop in properties where prop.type == FontFamilyProperty.name {
            touched = true
            cfg = FontFamilyConfig()
            // Case 1: array of entries — the common output of the parser.
            if case .array(let entries) = prop.data {
                for entry in entries {
                    // An entry can be an object {"name":"Helvetica"} or
                    // {"keyword":"serif"} or a bare string.
                    if case .object(let o) = entry {
                        if let n = o["name"]?.stringValue { cfg.names.append(n) }
                        if let kw = o["keyword"]?.stringValue {
                            classifyGeneric(kw, into: &cfg)
                        }
                    } else if let s = entry.stringValue {
                        // Bare string: either a name or a generic keyword.
                        classifyGeneric(s, into: &cfg)
                        if !isGeneric(s) { cfg.names.append(s) }
                    }
                }
            }
            // Case 2: single string blob.
            else if let s = prop.data.stringValue {
                classifyGeneric(s, into: &cfg)
                if !isGeneric(s) { cfg.names.append(s) }
            }
        }
        return touched ? cfg : nil
    }

    // Mark known generic family keywords so the applier can pick the right
    // SwiftUI Font.Design even when no real face name matches.
    private static func classifyGeneric(_ raw: String, into cfg: inout FontFamilyConfig) {
        switch raw.lowercased() {
        case "monospace", "ui-monospace": cfg.hasMonospace = true
        case "serif", "ui-serif":          cfg.hasSerif = true
        case "ui-rounded":                 cfg.hasRounded = true
        default: break   // other generics (cursive/fantasy/emoji) fall through
        }
    }

    // Returns true when `raw` is one of the CSS generic family keywords
    // and should not be added to `names` as a face name.
    private static func isGeneric(_ raw: String) -> Bool {
        switch raw.lowercased() {
        case "monospace", "serif", "sans-serif", "cursive", "fantasy",
             "system-ui", "ui-serif", "ui-sans-serif", "ui-monospace",
             "ui-rounded", "emoji", "math", "fangsong":
            return true
        default:
            return false
        }
    }
}

File: runtimes/web/src/engine/typography/FontFamilyExtractor.ts
// FontFamilyExtractor.ts — folds `FontFamily` IR properties into a FontFamilyConfig.
// Family: font-family.  IR shapes catalogued during Phase-6 survey of
// examples/properties/typography/*.json after `./gradlew run` conversion.

import { FontFamilyConfig, FONT_FAMILY_PROPERTY_TYPE, FontFamilyPropertyType } from './FontFamilyConfig';

// Minimal IRProperty shape — keeps engine modules decoupled from IR types dir.
interface IRPropertyLike { type: string; data: unknown; }

// Narrowing predicate; mirrors the other engine modules (Phase 4/5 pattern).
export function isFontFamilyProperty(type: string): type is FontFamilyPropertyType {
  return type === FONT_FAMILY_PROPERTY_TYPE;                                           // exact string match
}

// Per-family parse routine — returns the CSS value string (or undefined to drop).
function parse(data: unknown): string | number | undefined {
  // FontFamily: IR emits ["Serif","sans-serif"] (string[]).  Quote any family
  // whose name contains whitespace so the browser treats it as a single token;
  // generic families (serif/sans-serif/monospace/...) are never quoted.
  const GENERIC = new Set(['serif','sans-serif','monospace','cursive','fantasy',
    'system-ui','ui-serif','ui-sans-serif','ui-monospace','ui-rounded','emoji','math','fangsong']);
  const list: string[] = [];                                         // output accumulator
  const raw = Array.isArray(data)                                    // accept array or bare object
    ? data
    : (data && typeof data === 'object' && Array.isArray((data as Record<string, unknown>).families))
      ? (data as Record<string, unknown>).families as unknown[]
      : (typeof data === 'string' ? [data] : []);
  for (const f of raw) {                                             // iterate families
    if (typeof f !== 'string' || f.length === 0) continue;           // skip garbage
    const lower = f.toLowerCase();                                   // case-insensitive generic check
    if (GENERIC.has(lower)) list.push(lower);                        // emit generic unquoted
    else if (/[\s"']/.test(f)) list.push(`"${f.replace(/"/g,'\\"')}"`); // quote custom w/ space
    else list.push(f);                                               // single-token custom unquoted
  }
  return list.length ? list.join(', ') : undefined;                  // CSS comma-separated list
}

// Main entrypoint — last write wins, mirroring CSS cascade semantics.
export function extractFontFamily(properties: IRPropertyLike[]): FontFamilyConfig {
  const cfg: FontFamilyConfig = {};                                      // blank accumulator
  for (const p of properties) {
    if (!isFontFamilyProperty(p.type)) continue;                         // filter unrelated
    const v = parse(p.data);                                          // convert payload
    if (v !== undefined) cfg.value = v;                               // record result
  }
  return cfg;
}
